### DISCUSSION TITLE
How do I create a clock multiplier?

### DISCUSSION THREAD

**Discussion by me-nkr** [hneemann/Digital#1275] [Q&A]
I am trying to run an operation 16 times in one clock cycle. I believe I could achieve this by multiplying the base clock by 16. But how do I do it?

- I read online I can use delays to achieve this, but how do I do that? Like how do I use delays in general?

**Comment by TheWildJarvi** [comment]
Set the clock rate 16 times faster than you need for it then divide it down by 16 for the slower clock. On Apr 5, 2024 1:12 PM, Naveen K R ***@***.***> wrote:
I am trying to run an operation 16 times in one clock cycle. I believe I could achieve this by multiplying the base clock by 16. But how do I do it?
I read online I can use delays to achieve this, but how do I do that? Like how do I use delays in general?

**Reply by me-nkr** [reply]
Thanks for the suggestion, but I am specifically looking for a way to multiply clock speed since the circuit that uses the multiplied speed is an embedded circuit, and I need to pass the base clock as an input.

**Reply by TheWildJarvi** [reply]
AFAIK there is no way to multiply a clock in this simulator as this simulator is based on clock frequency defined by the clock component. In real life you could do something like a rising edge+ falling edge monostable to get 2 pulses out for every pulse input, but in this Simulator that will not work because of how the clocking is done. 

youre going to have to use a faster clock everywhere and divide it slower where you need it instead of a slower clock everywhere and multiply it faster where you need it.

**Reply by TheWildJarvi** [reply]
Also Delay components just delay the output by some amount of clock cycles from the input. this does not generate pulses it just delays when an output can happen. Delay components are not synthesizeable since they work in clock cycles in the simulator. The simulator has no idea what time is, only clock cycles.

**Comment by mengstr** [comment]
Something can be done with delay elements even in Digital. Though I'm not too sure I'd use this technique instead of the more reasonable "run-the-clock-at-16x-and-divide-the-clock-for-the-main-circuit"...
<img width="965" alt="Screenshot 2024-04-05 at 23 45 41" src="https://github.com/hneemann/Digital/assets/325326/394c64bb-ed30-4f7b-8f24-da56fe5019d6">

**Reply by TheWildJarvi** [reply]
this isnt a clock multiplier tho, this would be a clock divider still?

**Reply by mengstr** [reply]
Well. As seen on the screenshot the clock made 9 ticks and yet the output of the xor counted 54 ticks, so in some (maybe not in all though) regards it is multiplying the transitions of the  clock by a factor of 6. No dividing going on here at all.

Try the circuit. I set the clock at 1hz and the delay-elements at 10 cycles each. (It might work having two or even one as well)

**Comment by hneemann** [comment]
If delay elements are used, the generated clock signal depends on the runtime behavior of the gates used. This would not be done in reality, as there would be a very large temperature influence, for example. It could easily be that the multiplied signal "overtakes" the input clock, which could lead to interference.

If you wanted to generate a clock signal that is a multiple of an input clock in reality, you would use a PLL instead. This is done in every modern processor in order to generate the actual fast processor clock from the slow board clock by "multiplying" it by a fixed factor.

A PLL (phase-locked loop) is a control circuit consisting of a voltage-controlled oscillator and a frequency divider. The voltage and thus the frequency is increased until the divided clock signal is in phase with the input frequency. The oscillator frequency is then a corresponding multiple of the input frequency. In reality, therefore, a frequency division takes place.

In this simulator it is not possible to implement something like this because the simulator is event-based and there is no real notion of time. An "overtake" is not possible here, for example, as the clock signal would wait until the propagation times of the delay gates have elapsed.

**Reply by me-nkr** [reply]
So it's better to use a clock divider and not try spending time on a clock multiplier right?

- Also what is the duration in the delay component? is it seconds? nanoseconds or some other units?

**Reply by hneemann** [reply]
You could build something in the simulator that increases the frequency, but that would have little to do with reality and would be more about the properties of the simulator itself. From a didactic point of view, this makes little sense. At least in my opinion.

There is no fixed duration. The unit of delay is the duration of a single simulation step. The delay gate simply skips one simulation step. If there are many gates, the calculation takes longer, if there are few gates, it goes faster. The clock signal, for example, always waits until the calculation is complete and the circuit has stabilized. Therefore, the clock can never overtake a "delay gate multiplication", as it would be the case in reality.
